src/frontend/src/optimizer/mod.rs

```diff
@@ -294,8 +295,35 @@ impl PlanRoot {
             Convention::Logical => {
                 let plan = self.gen_optimized_logical_plan_for_stream()?;
 
-                let (plan, out_col_change) =
-                    plan.logical_rewrite_for_stream(&mut Default::default())?;
+                let (plan, out_col_change) = {
+                    let (plan, out_col_change) =
+                        plan.logical_rewrite_for_stream(&mut Default::default())?;
+                    if out_col_change.is_injective() {
+                        (plan, out_col_change)
+                    } else {
+                        let mut output_indices = (0..plan.schema().len()).collect_vec();
+                        #[allow(unused_assignments)]
+                        let (mut map, mut target_size) = out_col_change.into_parts();
+
+                        // TODO(st1page): https://github.com/risingwavelabs/risingwave/issues/7234
+                        // assert_eq!(target_size, output_indices.len());
+                        target_size = plan.schema().len();
+                        let mut tar_exists = vec![false; target_size];
+                        for i in map.iter_mut().flatten() {
+                            if tar_exists[*i] {
+                                output_indices.push(*i);
```

Comment:
basically, this won't work for the following query:

```sql
create table t(x int, y int);
explain create materialized view mv as  SELECT t.x x1, t.x x2, y FROM t join t tt ON t.x=tt.x;
````
As it will assign the name x2 to the y column and the name y to the secondary t.x column

Reply:
It's probably possible to work around so that the ordering is respected.

Reply:
> What bothers me a bit is that in our original problem, the element already exists somewhere in the output, we just have no idea where it is anymore, so we end up duplicating elements, and we no longer prune away the useless columns.

I am still much confused and do not know what's wrong... I think the root cause might be the `logical_rewrite_for_stream` of `LogicalJoin` adding the join key into the output_indices brutally to solve the "stream key issue" https://singularity-data.quip.com/3lUJAQMthect/RFC-Streamkey-instead-pk

Reply:
If the mapping is not injective, it will mistakenly believe that one of the keys is missing.

Reply:
Hmm indeed, the stream plan in https://github.com/risingwavelabs/risingwave/pull/8267 does indeed seem to contain the join key a second time erroneously, seems it does not completely get rid of repeated output indices (but in this case it does not cause non-injectivity) https://github.com/risingwavelabs/risingwave/blob/399ec7ca575f60c0034efcaeab89ed35fbdd5328/src/frontend/planner_test/tests/testdata/join.yaml#L612

Reply:
The logical_rewrite_for_stream's interface definition allows returning a non-injectivity change. And in fact the behavior gives the upper operator a chance to merge two columns into one. The issue might be we do not handle the non-injectivity here but not the join gives us a non-injectivity change?

Reply:
In other words, maybe making sure everyone calling the `logical_rewrite_for_stream` can handle the non-injectivity is easier than making sure every plannode's `logical_rewrite_for_stream` not returning a non-injectivity

Reply:
I still believe non injective is wrong. We end up becoming confused and forgetting certain columns and relying on their supposed aliases. I am afraid that something subtle will break due to this if we simply fix superficially. 

To be clear, duplicate output indices are not the problem. If we can rewrite `logical_rewrite_for_stream` so that it returns an injective mapping, there is nothing with having duplicate output indices

Reply:
> gives the upper operator a chance to merge two columns into one 

Perhaps, but we do not then prune away the aliased columns...

Reply:
> Perhaps, but we do not then prune away the aliased columns...

Yes. I mean that we cannot do the prune on the top of the stream plan or a Materialize. But how about other operators? considering if there is a Aggregation with max(input_ref(0)), max(input_ref(1))...

Reply:
> We end up becoming confused and forgetting certain columns and relying on their supposed aliases. 

In fact, this PR's fix gives the same plan with #8267. And actually, the query's output has two same columns, the only way is to copy it with a project operator.

Reply:
> the only way is to copy it with a project operator.

Well, what I'm saying is that the output actually has 3 of the same columns...

Reply:
Btw, why doesn't this work for stream?
```rust
LogicalProject::with_out_fields(plan, &self.out_fields)
```
I guess for stream, we still care about hidden columns?

Reply:
Yes, the required column is the columns declared in the user's query and the others is derived by `logical_rewrite_for_stream` and as the stream key.

Reply:
your fix also give 3 same columns...
```
- name: Repeated columns in project should not interfere with join result (https://github.com/risingwavelabs/risingwave/issues/8216)
  sql: |
    create table t(x int);
    SELECT t.x x1, t.x x2 FROM t join t tt ON t.x=tt.x;
  stream_plan: |
    StreamMaterialize { columns: [x1, x2, t._row_id(hidden), t._row_id#1(hidden), t.x(hidden)], pk_columns: [t._row_id, t._row_id#1, x2, t.x], pk_conflict: "no check" }
    └─StreamProject { exprs: [t.x, t.x, t._row_id, t._row_id, t.x] }
      └─StreamHashJoin { type: Inner, predicate: t.x = t.x, output: [t.x, t._row_id, t._row_id, t.x] }
        ├─StreamExchange { dist: HashShard(t.x) }
        | └─StreamTableScan { table: t, columns: [t.x, t._row_id], pk: [t._row_id], dist: UpstreamHashShard(t._row_id) }
        └─StreamExchange { dist: HashShard(t.x) }
          └─StreamTableScan { table: t, columns: [t.x, t._row_id], pk: [t._row_id], dist: UpstreamHashShard(t._row_id) }

```

Reply:
the two non-hidden column is produced by the non-injectivity returning which is unavoidable. And the hidden column is produced by the `LogicalJoin::logical_rewrite_for_stream`. The only thing we can optimize is to check if the join key has been in the original output indices in the rewriting. But we should do nothing for the non-injectivity. They are totally two different things and we can not mix them up.

Reply:
This is inaccurate. This PR produces 4 of the same column:
```
[t.x(hidden), x1, t._row_id(hidden), t._row_id#1(hidden), t.x#1(hidden), x2]
```
at pos 0, 1, and -1, -2

Reply:
So the join's implementation has some wrong and should reduce the column he adds. But the issue is not related to the non-injectivity.